### ISSUE TITLE
MNIST VAE Documentation

### ISSUE THREAD

**Issue by GeorgeNash** [tensorflow/probability#56]
Hi team,

You've done some great work creating Jupyter notebooks for the `Eight Schools` problem etc.

I'm working through your example scripts, in particular: https://github.com/tensorflow/probability/blob/master/tensorflow_probability/examples/vae.py

However, it's quite tough to grapple what's going on inside the VAE from the code comments alone and I'm struggling to decipher the large amount of output that's being produced. The questions I'm looking to ask are: 

- what's happening inside the VAE at each timestep
- what the output images at each time step signify
- what's a `good` value for my loss
- the expected result

Do you have any existing documentation that may cover this? Even a small amount of documentation would greatly help me here. 

If you don't, that's fine, perhaps I can help to produce some?

Thanks,
George

**Comment by GeorgeNash** [comment]
For example, I am running 10,000 steps and my current MNIST reconstructions (at step 10,000) look like the following:

`Step: 9900 Loss: 13249.059 (0.008 sec)`

![step09999_validation_reconstructions](https://user-images.githubusercontent.com/5854544/40663700-2dc1bcd8-6351-11e8-8bed-70c7961540c9.png)

These reconstructions seem way off.

It would be great if you could document some expected output and suggested parameters so I would know if I was on the right track

**Comment by davmre** [comment]
Hi George, thanks for reporting this! It turns out this bug was introduced in migrating the example from `tf.layers` to `tf.keras.layers` -- the training is actually working fine under the hood, but the visualized reconstructions are random noise because Keras uses a different approach to variable scoping, so it fails to share weights between the training code path and the visualization code path.

For now we've reverted to the older approach, so it should work now if you try running the example from github [HEAD](https://github.com/tensorflow/probability/commit/b6b70dbc4368b045fe42fea64510d1a415d8bac1) (or the `tfp-nightly` package after tonight).

Sorry for the confusion! Let us know if you run into more issues.
